RIOT-OS/RIOT
# Add some Rust building infrastructure and example

### Contribution description

While writing Rust applications has been possible for some time, it has required a few Makefile rules which I'm by now happy enough with to start a first PR. Having this all in-tree is not essential for operation, but makes it a bit easier to store, for example, the target definitions with the architectures rather than [in a big if chain](https://gitlab.com/etonomy/riot-examples/-/blob/master/Makefile.cargo-target) that sometimes peeks into fields and sometimes into boards.

This adds
* Rust target definitions to platforms -- basically Rust's version of a platform triple -- in the architectures
* Build rules to create something almost, bot not quite, entirely unlike modules -- it's called `$(APPLICATION_RUST_MODULE).module` and added to `BASELIBS`, but is built by running Cargo and then exploding the built static lib into the folder where LINK expects it to be.
* Two examples

### Testing procedure

* Ensure you have Rust installed -- given you need nightly versions, typically by installing [rustup](https://rustup.rs/)
  * Alternatively, run with `BUILD_IN_DOCKER=1`
* In examples/rust-hello-world (or rust-gcoap), run `make all flash term`, optionally with a cortex M0-4 board.

It is tested so far with native and some cortex-m0, 3 and 4 boards (mostly EFM32, EFR32 and nRF52840). riscv support is disabled right now, I don't want to do that in the first rush until it falls into place trivially.


### Questions:

* Did I get the what-goes-where of the Makefiles right?
* Would it make more sense to build something parallel to `${MODULE}.module` that is `${CRATE}.crate` that gets its own place in the big find of LINK? I think it'd make sense because it wouldn't involve overriding a target any more, and moreover make it easier to later extend this to having potentially multiple built crates (although I think it's preferable to use one), but would make a larger diff.
  * If that's solved, I can probably come up with a nicer way of saying, in the example, that "this has a crate as its main rather than a module", and show less guts in there.

### Issues/PRs references

Closes: https://github.com/RIOT-OS/RIOT/issues/9799

[edit: Updated examples list and test description now that the riotdocker images just do the right thing]
